itests/deals_concurrent_test.go

```diff
@@ -107,6 +111,79 @@ func TestDealCyclesConcurrent(t *testing.T) {
 	}
 }
 
+// TestDealCycleProviderRestart verifies that
+// - the client makes a deal with the storage provider
+// - the storage provider restarts
+// - the client can make a retrieval deal with the storage provider
+func TestDealCycleProviderRestart(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode")
+	}
+
+	kit.QuietMiningLogs()
+
+	blockTime := 10 * time.Millisecond
+
+	// For these tests where the block time is artificially short, just use
+	// a deal start epoch that is guaranteed to be far enough in the future
+	// so that the deal starts sealing in time
+	startEpoch := abi.ChainEpoch(2 << 12)
+
+	// To simulate restarting the provider, we create two storage providers,
+	// and share a DAG store between them.
+	// We make a storage deal against the first provider, then start the second
+	// provider with the same DAG store and make a retrieval deal aaginst the
+	// second provider.
+	var (
+		client    kit.TestFullNode
+		provider1 kit.TestMiner
+		provider2 kit.TestMiner
+	)
+
+	// Use a single DAG store shared between miner1 and miner2
+	dagStoreDir := t.TempDir()
+	dagStoreDS := ds_sync.MutexWrap(ds.NewMapDatastore())
+	dagStore, err := dagstore.NewDAGStore(dagstore.Config{
+		TransientsDir: filepath.Join(dagStoreDir, "transients"),
+		IndexDir:      filepath.Join(dagStoreDir, "index"),
+		Datastore:     dagStoreDS,
+	})
+	require.NoError(t, err)
+	consOpts := kit.ConstructorOpts(
+		node.Override(new(*dagstore.DAGStore), dagStore),
+	)
+
+	// Set up the first provider
+	opts := []kit.NodeOpt{kit.WithAllSubsystems(), consOpts}
+	ens := kit.NewEnsemble(t, kit.MockProofs()).
+		FullNode(&client, opts...).
+		Miner(&provider1, &client, opts...).
+		Start()
+
+	ens.InterconnectAll().BeginMining(blockTime, &provider1)
+	dh := kit.NewDealHarness(t, &client, &provider1, &provider1)
+
+	// Make a storage deal against the first provider
+	deal, res, inPath := dh.MakeOnlineDeal(context.Background(), kit.MakeFullDealParams{
+		Rseed:      5,
+		StartEpoch: startEpoch,
+	})
+
+	// Set up the second provider
+	provider2Opts := append(opts,
+		kit.OwnerAddr(client.DefaultKey),
+		kit.MainMiner(&provider1),
```

Comment:
@aarshkshah1992 this doesn't make sense to me. AFAICT you want to create a second provider with the same DAG store, but this is not what this test is doing. Here we are creating a second provider and linking it to the first provider, as if you have a market process and a miner process (this is what the `MainMiner` option is for).

---

I think the idea is OK, why don't you just start a new ensemble, same as you started the first one, and just link to the same DAG store?

Reply:
Also I don't see us stopping the initial miner - not sure if we want to, or not, but I think we want to.

Reply:
I tried following the suggestion of creating a second ensemble but unfortunately this is complicated to do because there are multiple pieces that need to be shared in order to perform a full deal cycle, for example the piece store.
I think the correct solution is what Anton suggests - we need to be able to stop and start nodes.

I'm going to close this for now and we can come back to it when it's possible to stop and start nodes.